File: dotnet/src/webdriver/WebDriverCredentials.cs
// <copyright file="WebDriverCredentials.cs" company="WebDriver Committers">
// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements. See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership. The SFC licenses this file
// to you under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// </copyright>

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Text;
using Newtonsoft.Json;

namespace OpenQA.Selenium
{
    /// <summary>
    /// Describes the type of authentication.
    /// </summary>
    public enum AuthType
    {
        /// <summary>
        ///  Normal HTTP authentication
        /// </summary>
        Normal,

        /// <summary>
        /// NTLM ("Windows") authentication
        /// </summary>
        NTLM,

        /// <summary>
        /// No explicit authentication
        /// </summary>
        Unspecified
    }
    
    /// <summary>
    /// Describes authentication settings to be used with a driver instance.
    /// </summary>
    [JsonObject(MemberSerialization.OptIn)]
    public class WebDriverCredentials
    {
        private AuthType authType = AuthType.Unspecified;
        private String username;
        private String password;
        private String host;
        private int port = -1;
        private String realm;
        private String ntlmWorkstation;
        private String ntlmDomain;

        /// <summary>
        /// Initializes a new instance of the <see cref="WebDriverCredentials"/> class.
        /// </summary>
        public WebDriverCredentials()
        {
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="WebDriverCredentials"/> class with the given authentication settings.
        /// </summary>
        /// <param name="settings">A dictionary of settings to use for authentication.</param>
        public WebDriverCredentials(Dictionary<string, object> settings)
        {
            if (settings == null)
            {
                throw new ArgumentNullException("settings", "settings dictionary cannot be null");
            }

            if (settings.ContainsKey("authType"))
            {
                AuthType rawType = (AuthType)Enum.Parse(typeof(AuthType), settings["authType"].ToString(), true);
                this.Type = rawType;
            }

            if (settings.ContainsKey("username"))
            {
                this.Username = settings["username"].ToString();
            }
            
            if (settings.ContainsKey("password"))
            {
                this.Password = settings["password"].ToString();
            }
            
            if (settings.ContainsKey("host"))
            {
                this.Host = settings["host"].ToString();
            }
            
            if (settings.ContainsKey("port"))
            {
                if (settings["port"] is int) {
                    this.Port = (int)settings["port"];
                }
                else if(!int.TryParse(settings["port"].ToString(), out this.port)) {
                    this.port = -1;
                }
            }
            
            if (settings.ContainsKey("realm"))
            {
                this.Realm = settings["realm"].ToString();
            }

            if (settings.ContainsKey("ntlmWorkstation"))
            {
                this.NtlmWorkstation = settings["ntlmWorkstation"].ToString();
            }

            if (settings.ContainsKey("ntlmDomain"))
            {
                this.NtlmDomain = settings["ntlmDomain"].ToString();
            }
        }

        /// <summary>
        /// Gets or sets the type of authentication.
        /// </summary>
        [JsonIgnore]
        public AuthType Type
        {
            get 
            {
                return this.authType; 
            }

            set
            {
                this.authType = value;
            }
        }

        /// <summary>
        /// Gets the type of authentication as a string for JSON serialization.
        /// </summary>
        [JsonProperty("authType")]
        public string SerializableProxyKind {
            get {
                return this.authType.ToString().ToUpperInvariant();
            }
        }

        /// <summary>
        /// Gets or sets the value of the username for authentication.
        /// </summary>
        [JsonProperty("username", DefaultValueHandling = DefaultValueHandling.Ignore, NullValueHandling = NullValueHandling.Ignore)]
        public string Username {
            get {
                return this.username;
            }

            set {
                this.username = value;
            }
        }

        /// <summary>
        /// Gets or sets the value of the password for authentication.
        /// </summary>
        [JsonProperty("password", DefaultValueHandling = DefaultValueHandling.Ignore, NullValueHandling = NullValueHandling.Ignore)]
        public string Password {
            get {
                return this.password;
            }

            set {
                this.password = value;
            }
        }

        /// <summary>
        /// Gets or sets the host to which the credentials apply.
        /// </summary>
        [JsonProperty("host", DefaultValueHandling = DefaultValueHandling.Ignore, NullValueHandling = NullValueHandling.Ignore)]
        public string Host {
            get {
                return this.host;
            }

            set {
                this.host = value;
            }
        }

        /// <summary>
        /// Gets or sets the port to which the credentials apply.
        /// </summary>
        [JsonProperty("port", DefaultValueHandling = DefaultValueHandling.Ignore, NullValueHandling = NullValueHandling.Ignore)]
        public int Port {
            get {
                return this.port;
            }

            set {
                this.port = value;
            }
        }

        /// <summary>
        /// Gets or sets the realm to which the credentials apply.
        /// </summary>
        [JsonProperty("realm", DefaultValueHandling = DefaultValueHandling.Ignore, NullValueHandling = NullValueHandling.Ignore)]
        public string Realm {
            get {
                return this.realm;
            }

            set {
                this.realm = value;
            }
        }

        /// <summary>
        /// Gets or sets the workstation for NTLM authentication.
        /// </summary>
        [JsonProperty("ntlmWorkstation", DefaultValueHandling = DefaultValueHandling.Ignore, NullValueHandling = NullValueHandling.Ignore)]
        public string NtlmWorkstation {
            get {
                return this.ntlmWorkstation;
            }

            set {
                this.ntlmWorkstation = value;
            }
        }

        /// <summary>
        /// Gets or sets the domain for NTLM authentication.
        /// </summary>
        [JsonProperty("ntlmDomain", DefaultValueHandling = DefaultValueHandling.Ignore, NullValueHandling = NullValueHandling.Ignore)]
        public string NtlmDomain {
            get {
                return this.ntlmDomain;
            }

            set {
                this.ntlmDomain = value;
            }
        }
    }
}

File: java/client/src/org/openqa/selenium/WebDriverCredentials.java
// Licensed to the Software Freedom Conservancy (SFC) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The SFC licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

package org.openqa.selenium;

import java.util.Map;

/**
 * Configuration parameters for using credentials in WebDriver. Drivers which support these
 * settings will use the specified credentials by default.
 */
public class WebDriverCredentials {

  public enum AuthType {

	  /** Normal HTTP authentication */
	  NORMAL,
	  /** NTLM ("Windows") authentication */
	  NTLM,
	  /** No explicit authentication */
	  UNSPECIFIED
  }

  private AuthType authType = AuthType.UNSPECIFIED;
  private String username;
  private String password;
  private String host;
  private int port = -1;
  private String realm;
  private String ntlmWorkstation;
  private String ntlmDomain;

  public WebDriverCredentials() {
    // Empty default constructor
  }

  public WebDriverCredentials(Map<String, ?> raw) {
    if (raw.containsKey("authType") && raw.get("authType") != null) {
      setAuthType(AuthType.valueOf(((String) raw.get("authType")).toUpperCase()));
    }
    if (raw.containsKey("username") && raw.get("username") != null) {
      setUsername((String) raw.get("username"));
    }
    if (raw.containsKey("password") && raw.get("password") != null) {
      setPassword((String) raw.get("password"));
    }
    if (raw.containsKey("host") && raw.get("host") != null) {
      setHost((String) raw.get("host"));
    }
    if (raw.containsKey("port") && raw.get("port") != null) {
      setPort(((Number) raw.get("port")).intValue());
    }
    if (raw.containsKey("realm") && raw.get("realm") != null) {
      setRealm((String) raw.get("realm"));
    }
    if (raw.containsKey("ntlmWorkstation") && raw.get("ntlmWorkstation") != null) {
      setNtlmWorkstation((String) raw.get("ntlmWorkstation"));
    }
    if (raw.containsKey("ntlmDomain") && raw.get("ntlmDomain") != null) {
      setNtlmDomain((String) raw.get("ntlmDomain"));
    }
  }
  
  /**
   * Gets the {@link AuthType}. 
   * It defaults to {@link AuthType#UNSPECIFIED}.
   *
   * @return the authentication type employed
   */
  public AuthType getAuthType() {
	  return this.authType;
  }
  
  /**
   * Explicitly sets the authentication type
   * 
   * @return reference to self
   */
  public WebDriverCredentials setAuthType(AuthType authType) {
	  this.authType = authType;
	  return this;
  }
  
  /**
   * Gets the username used for authentication.
   * 
   * @return the username if present, or null if not set
   */
  public String getUsername() {
	  return username;
  }
  
  /**
   * Specify which username to use for authentication.
   * 
   * @param username
   * @return reference to self
   */
  public WebDriverCredentials setUsername(String username) {
	  this.username = username;
	  return this;
  }
  
  /**
   * Gets the authentication password.
   * 
   * @return the authentication password if present, or null if not set
   */
  public String getPassword() {
	  return password;
  }
  
  /**
   * Specify which password to use for authentication.
   * 
   * @param password
   * @return reference to self
   */
  public WebDriverCredentials setPassword(String password) {
	  this.password = password;
	  return this;
  }
  
  /**
   * Gets authentication host to which the credentials apply
   * 
   * @return The authentication host, or null if not set
   */
  public String getHost() {
	  return host;
  }
  
  /**
   * Specify which port to which the credentials apply
   * 
   * @param host
   * @return reference to self
   */
  public WebDriverCredentials setHost(String host) {
	  this.host = host;
	  return this;
  }
  
  /**
   * Gets the port to which the credentials apply.
   * 
   * @return the port, or -1 if unspecified
   */
  public int getPort() {
	  return port;
  }

  /**
   * Specify which port to which the credentials apply
   * @param port The port; use -1 for unspecified
   * @return reference to self
   */
  public WebDriverCredentials setPort(int port) {
	  this.port = port;
	  return this;
  }
  
  /**
   * Gets the realm to which the credentials apply
   * @return The realm, or null if not set
   */
  public String getRealm() {
	  return realm;
  }
  
  /**
   * Specify the realm to which the credentials apply.
   * @param realm
   * @return reference to self
   */
  public WebDriverCredentials setRealm(String realm) {
	  this.realm = realm;
	  return this;
  }
  
  /**
   * Gets the workstation for NTLM authentication
   * @return the workstation
   */
  public String getNtlmWorkstation() {
	  return ntlmWorkstation;
  }
  
  /**
   * Specifies the workstation for NTLM authentication
   * 
   * @param ntlmWorkstation The workstation
   * @return reference to self
   */
  public WebDriverCredentials setNtlmWorkstation(String ntlmWorkstation) {
	  this.ntlmWorkstation = ntlmWorkstation;
	  return this;
  }
  
  /**
   * Gets the domain for NTLM authentication
   * 
   * @return the domain
   */
  public String getNtlmDomain() {
	  return ntlmDomain;
  }
  
  /**
   * Specifies the domain for NTLM authentication
   * 
   * @param ntlmDomain The domain
   * @return reference to self
   */
  public WebDriverCredentials setNtlmDomain(String ntlmDomain) {
	  this.ntlmDomain = ntlmDomain;
	  return this;
  }

  @SuppressWarnings({"unchecked"})
  public static WebDriverCredentials extractFrom(Capabilities capabilities) {
    Object rawCredentials = capabilities.getCapability("credentials");
    WebDriverCredentials credentials = null;
    if (rawCredentials != null) {
      if (rawCredentials instanceof WebDriverCredentials) {
    	  credentials = (WebDriverCredentials) rawCredentials;
      } else if (rawCredentials instanceof Map) {
    	  credentials = new WebDriverCredentials((Map<String, ?>) rawCredentials);
      }
    }
    return credentials;
  }
}
